### PR TITLE
Direct dispatch to fleet or robot

### PR DESCRIPTION
**Pull request by aaronchongth**
## What's new

* Resolves https://github.com/open-rmf/rmf-web/issues/636
* Support fleet and robot direct dispatches

image
image

https://github.com/user-attachments/assets/d639c49c-ad5c-43aa-a594-cf7c2f12226b

## Self-checks

- [ ] I have prototyped this new feature (if necessary) on Figma 
- [ ] I'm familiar with and follow this [Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [ ] I added unit-tests for new components
- [ ] I tried testing edge cases
- [ ] I tested the behavior of the components that interact with the backend, with an e2e test

### CODE REVIEW

File: packages/react-components/lib/tasks/create-task.tsx [hunk 1]
```diff
@@ -930,6 +989,90 @@ export function CreateTaskForm({
...
+                        fullWidth
+                        margin="normal"
+                      />
+                    )}
+                  </Grid>
+                </Grid>
+                <Divider
+                  orientation="horizontal"
+                  flexItem
+                  style={{ marginTop: theme.spacing(2), marginBottom: theme.spacing(2) }}
```
Comment (koonpeng) [inline packages/react-components/lib/tasks/create-task.tsx:1074]: Use `sx` instead of `style`

**Review by koonpeng — CHANGES_REQUESTED** [review]

File: packages/react-components/lib/tasks/create-task.tsx (hunk 1 above)
Comment (aaronchongth) [reply packages/react-components/lib/tasks/create-task.tsx]: https://github.com/open-rmf/rmf-web/pull/1004/commits/cdf7fbaef4c5ee34d37ba2fce715425412da2c39

File: packages/react-components/lib/tasks/create-task.tsx [hunk 2]
```diff
@@ -930,6 +989,90 @@ export function CreateTaskForm({
                   flexItem
                   style={{ marginTop: theme.spacing(2), marginBottom: theme.spacing(2) }}
                 />
+                <Grid container spacing={theme.spacing(2)} alignItems="center">
+                  <Grid item xs={6}>
+                    <TextField
+                      select
+                      id="dispatch-type"
+                      label="Dispatch type"
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/react-components/lib/tasks/create-task.tsx]: https://github.com/open-rmf/rmf-web/pull/1004/commits/cdf7fbaef4c5ee34d37ba2fce715425412da2c39

File: packages/dashboard/src/hooks/useCreateTaskForm.tsx [hunk 3]
```diff
@@ -42,9 +43,21 @@ export const useCreateTaskFormData = (rmf: RmfIngress | undefined) => {
...
+    subs.push(
+      rmf.fleetsObs.subscribe((fleetStates) => {
+        const result: Record<string, string[]> = {};
+        for (const fleet of fleetStates) {
+          if (!fleet.name || !fleet.robots) {
+            continue;
+          }
+          result[fleet.name] = Object.keys(fleet.robots);
+        }
+        setFleets(result);
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/hooks/useCreateTaskForm.tsx:55]: This is currently used within a single `useCreateTaskFormData` call when `appbar` is loaded, so it only ever triggers a render once. This can be verified with running the hotel world.

File: packages/react-components/lib/tasks/create-task.tsx [hunk 4]
```diff
@@ -803,6 +812,56 @@ export function CreateTaskForm({
...
+    }
+    let robotFleet: string | null = null;
+    for (const fleetName of Object.keys(fleets)) {
+      if (fleets[fleetName].includes(ev.target.value)) {
+        robotFleet = fleetName;
+        break;
+      }
+    }
+    if (robotFleet === null) {
+      console.error(`Failed to find fleet name for robot [${ev.target.value}]`);
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/react-components/lib/tasks/create-task.tsx:863]: The text field is not an auto-complete, therefore users can only ever select what is available, I added a comment to indicate in that section of code that it is highly unlikely this will happen

https://github.com/open-rmf/rmf-web/pull/1004/commits/35f92a8f12b85096b1f30f19674cabd396413434

File: packages/dashboard/src/components/appbar.tsx [hunk 5]
```diff
@@ -194,38 +194,49 @@ export const AppBar = React.memo(({ extraToolbarItems }: AppBarProps): React.Rea
...
+        console.debug(robotTask);
+        await rmf.tasksApi.postRobotTaskTasksRobotTaskPost(robotTask);
       } else {
-        const scheduleRequests = taskRequests.map((req) => {
-          console.debug('schedule task:');
-          console.debug(req);
-          console.debug(schedule);
-          return toApiSchedule(req, schedule);
+        console.debug('dispatch task:');
+        console.debug(taskRequest);
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/appbar.tsx]: https://github.com/open-rmf/rmf-web/pull/1004/commits/35f92a8f12b85096b1f30f19674cabd396413434

File: packages/dashboard/src/components/appbar.tsx [hunk 6]
```diff
@@ -194,38 +194,49 @@ export const AppBar = React.memo(({ extraToolbarItems }: AppBarProps): React.Rea
...
   );
 
+  const scheduleTask = React.useCallback<Required<CreateTaskFormProps>['scheduleTask']>(
+    async (taskRequest, schedule) => {
+      if (!rmf) {
+        throw new Error('tasks api not available');
+      }
+      console.debug('schedule task:');
+      console.debug(taskRequest);
+      console.debug(schedule);
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/appbar.tsx]: https://github.com/open-rmf/rmf-web/pull/1004/commits/35f92a8f12b85096b1f30f19674cabd396413434

File: packages/dashboard/src/components/appbar.tsx [hunk 7]
```diff
@@ -194,38 +194,49 @@ export const AppBar = React.memo(({ extraToolbarItems }: AppBarProps): React.Rea
...
-        );
+      if (robotDispatchTarget) {
+        const robotTask: RobotTaskRequest = {
+          type: 'robot_task_request',
+          robot: robotDispatchTarget.robot,
+          fleet: robotDispatchTarget.fleet,
+          request: taskRequest,
+        };
+        console.debug(`dispatch robot task:`);
+        console.debug(robotTask);
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/appbar.tsx]: https://github.com/open-rmf/rmf-web/pull/1004/commits/35f92a8f12b85096b1f30f19674cabd396413434

File: packages/react-components/lib/tasks/create-task.tsx [hunk 8]
```diff
@@ -803,6 +812,56 @@ export function CreateTaskForm({
     });
   };
 
+  enum DispatchType {
+    Automatic = 'Automatic',
+    Fleet = 'Fleet',
+    Robot = 'Robot',
+  }
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/react-components/lib/tasks/create-task.tsx:819]: Hmm, I would rather keep these values close to where it is used to help with readability, unless it is needed elsewhere.

Just found out online that typescript will optimize out `const enum` and just use string literals, https://www.typescriptlang.org/play/?#code/PTAEGcFMCcDcEsDGlzEdAngBwC4HsA6HcAKETwDtwdRIKBXAW1AGFNc8X7po7ENQAbxKhRoAEQAhACotxoALwSZcgDQix4gKLSAEvKXa949WInSA8gDkWFgJJWD567YcmNo8QGULAGSfefu5m4gAKQYoS4f6mmtIASgAaAQnJJAC+JGSU4HgANpCgBHl4AOYAFGzY+Fw8fBgEqQCUQA

File: packages/react-components/lib/tasks/create-task.tsx [hunk 9]
```diff
@@ -930,6 +989,90 @@ export function CreateTaskForm({
...
+                        defaultValue=""
+                        onChange={handleDispatchRobotTargetChange}
+                      >
+                        {fleets &&
+                          Object.keys(fleets).flatMap((fleetName) => {
+                            const fleetRobots = [
+                              <Divider key={`${fleetName}_divider`} />,
+                              <MenuItem value={fleetName} key={fleetName} disabled>
+                                {fleetName}
+                              </MenuItem>,
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/react-components/lib/tasks/create-task.tsx:1053]: image

File: packages/react-components/lib/tasks/create-task.tsx [hunk 10]
```diff
@@ -930,6 +989,90 @@ export function CreateTaskForm({
...
+                          })}
+                      </TextField>
+                    ) : (
+                      <TextField
+                        disabled
+                        id="dispatch-automatic-type"
+                        variant="outlined"
+                        fullWidth
+                        margin="normal"
+                      />
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/react-components/lib/tasks/create-task.tsx]: https://github.com/open-rmf/rmf-web/pull/1004/commits/d9c8fbabe1c7f08bf4ba614bcd7bcfeb0d294e23

File: packages/dashboard/src/hooks/useCreateTaskForm.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (koonpeng) [reply packages/dashboard/src/hooks/useCreateTaskForm.tsx:55]: Tried this out, so it doesn't re-render because the api-client has a "fake" fleet observable, it only ever emits once and doesn't update when the fleet changes. Not sure if that is a bug but it's become a "feature" now, so probably just add a comment at api-client.

https://github.com/open-rmf/rmf-web/blob/d9c8fbabe1c7f08bf4ba614bcd7bcfeb0d294e23/packages/dashboard/src/services/rmf-ingress.ts#L226-L231

[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/hooks/useCreateTaskForm.tsx:55]: Good catch, thanks for pointing it out. https://github.com/open-rmf/rmf-web/pull/1004/commits/7ba206421c5899f57c5c38f92e8c96e8fe785a7f

File: packages/dashboard/src/components/tasks/task-schedule.tsx [hunk 11]
```diff
@@ -192,17 +196,45 @@ export const TaskSchedule = () => {
...
+        console.debug(taskRequest);
+        await rmf.tasksApi.postDispatchTaskTasksDispatchTaskPost({
+          type: 'dispatch_task_request',
+          request: taskRequest,
+        });
+      }
+      AppEvents.refreshTaskApp.next();
+    },
+    [rmf],
+  );
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/tasks/task-schedule.tsx]: https://github.com/open-rmf/rmf-web/pull/1004/commits/7ba206421c5899f57c5c38f92e8c96e8fe785a7f

File: packages/dashboard/src/components/tasks/task-schedule.tsx [hunk 12]
```diff
@@ -337,7 +366,8 @@ export const TaskSchedule = () => {
             setEventScope(EventScopes.CURRENT);
             AppEvents.refreshTaskSchedule.next();
           }}
-          submitTasks={submitTasks}
+          dispatchTask={dispatchTask}
+          scheduleTask={scheduleTask}
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply packages/dashboard/src/components/tasks/task-schedule.tsx]: they do very different things, so I would keep them separate but with more specific actions

File: packages/dashboard/src/components/tasks/task-schedule.tsx [hunk 13]
```diff
@@ -331,7 +350,8 @@ export const TaskSchedule = () => {
             setEventScope(EventScopes.CURRENT);
             AppEvents.refreshTaskSchedule.next();
           }}
-          submitTasks={submitTasks}
+          onDispatchTask={dispatchTaskCallback}
+          onScheduleTask={editScheduledTaskCallback}
```
Comment (koonpeng) [inline packages/dashboard/src/components/tasks/task-schedule.tsx:354]: Can we use the same code for both this and the appbar? The way it is now has quite abit of code smell, `onScheduleTask` has different meaning depending on the context, we need to decouple it, my ideas are either

1. Have 2 different callback explicitly, for create and edit, then both here and appbar can use the same code.
2. Use different components, i.e. have a `EditTask` component, it can use `CreateTaskForm` internally but with some differences in behavior and text.

**Review by koonpeng — CHANGES_REQUESTED** [review]

File: packages/dashboard/src/components/tasks/task-schedule.tsx (hunk 13 above)
Comment (aaronchongth) [reply packages/dashboard/src/components/tasks/task-schedule.tsx]: https://github.com/open-rmf/rmf-web/pull/1004/commits/4dece1e03f8d894d0a46602f1b0129e71b55fd68

File: packages/react-components/lib/tasks/task-form.tsx [hunk 14]
```diff
@@ -311,11 +299,14 @@ export interface CreateTaskFormProps
...
-  onDispatchTask(task: TaskRequest, robotDispatchTarget: RobotDispatchTarget | null): Promise<void>;
-  onScheduleTask(task: TaskRequest, schedule: Schedule): Promise<void>;
+  schedule?: Schedule;
+  taskRequest?: TaskRequest;
+  onDispatchTask?(
+    task: TaskRequest,
+    robotDispatchTarget: RobotDispatchTarget | null,
+  ): Promise<void>;
+  onScheduleTask?(task: TaskRequest, schedule: Schedule): Promise<void>;
+  onEditScheduleTask?(task: TaskRequest, schedule: Schedule): Promise<void>;
```
Comment (koonpeng) [inline packages/react-components/lib/tasks/task-form.tsx:309]: Add docs for these 2 props and `schedule`. iiuc, only `onEditScheduleTask` is used in "edit" mode, which is controlled by the `schedule` prop and `onScheduleTask` is used otherwise. They are mutually exclusive and will never be used together.

Better would be to refactor this to use the micro app hooks, the task form should be able to handle all the api calls itself and all these callbacks should be able to be replaced with a single `onClose` callback. But I leave it up to you if you want to do the refactor now.

**Review by koonpeng — APPROVED** [review]

File: packages/react-components/lib/tasks/task-form.tsx (hunk 14 above)
Comment (aaronchongth) [reply packages/react-components/lib/tasks/task-form.tsx:309]: they technically can be used together, just that we choose not to provide a callback in whichever place we are creating a task form, i.e. the edit schedule component can also provide `onScheduleTask` but because of the function itself, we choose not to provide it.

yeah using hooks will be ideal, I will help look into it when we finally merge react components and dashboard
